pkg/resource/bucket/custom_acl_code.go

```diff
@@ -0,0 +1,216 @@
+)
+
+var (
+	HEADER_USER_ID_FORMAT = "id=%s"
+	HEADER_URI_FORMAT     = "uri=%s"
+)
+
+type aclGrantHeaders struct {
+	FullControl string
+	Read        string
+	ReadACP     string
+	Write       string
+	WriteACP    string
+}
+
+// hasOwnerFullControl returns true if any of the grants matches the owner
+// and has full control permissions.
+func hasOwnerFullControl(owner *svcsdk.Owner, grants []*svcsdk.Grant) bool {
+	for _, grant := range grants {
+		if grant.Grantee == nil ||
+			grant.Grantee.ID == nil ||
+			*grant.Grantee.ID != *owner.ID {
+			continue
+		}
+
+		return *grant.Permission == svcsdk.PermissionFullControl
+	}
+	return false
+}
+
+// grantsContainPermission will return true if any of the grants have the
+// permission matching the one supplied.
+func grantsContainPermission(permission string, grants []*svcsdk.Grant) bool {
+	for _, grant := range grants {
+		if *grant.Permission == permission {
+			return true
+		}
+	}
+	return false
+}
+
+// getGrantsByGroupURI searches a list of ACL grants for any that have a
+// group type grantee with the given URI.
+func getGrantsByGroupURI(uri string, grants []*svcsdk.Grant) []*svcsdk.Grant {
+	matching := []*svcsdk.Grant{}
+
+	for _, grant := range grants {
+		if grant.Grantee == nil {
+			continue
+		}
+
+		if *grant.Grantee.Type != svcsdk.TypeGroup {
+			continue
+		}
+
+		if *grant.Grantee.URI == uri {
+			matching = append(matching, grant)
+		}
+	}
+	return matching
+}
+
+// getGrantsByCanonicalUserID searches a list of ACL grants for any that have a
+// canonical user type grantee with the given ID.
+func getGrantsByCanonicalUserID(id string, grants []*svcsdk.Grant) []*svcsdk.Grant {
+	matching := []*svcsdk.Grant{}
+
+	for _, grant := range grants {
+		if grant.Grantee == nil {
+			continue
+		}
+
+		if *grant.Grantee.Type != svcsdk.TypeCanonicalUser {
+			continue
+		}
+
+		if *grant.Grantee.ID == id {
+			matching = append(matching, grant)
+		}
+	}
+	return matching
+}
+
+// getGrantsByPermission searches a list of ACL grants for any that have the
+// given permission.
+func getGrantsByPermission(permission string, grants []*svcsdk.Grant) []*svcsdk.Grant {
+	matching := []*svcsdk.Grant{}
+
+	for _, grant := range grants {
+		if *grant.Permission == permission {
+			matching = append(matching, grant)
+		}
+	}
+	return matching
+}
+
+// formGrantHeader will form a grant header string from a list of grants
+func formGrantHeader(grants []*svcsdk.Grant) string {
+	headers := []string{}
+	for _, grant := range grants {
+		if grant.Grantee == nil {
+			continue
+		}
+
+		if *grant.Grantee.Type == svcsdk.TypeGroup {
+			headers = append(headers, fmt.Sprintf(HEADER_URI_FORMAT, *grant.Grantee.URI))
+		}
+		if *grant.Grantee.Type == svcsdk.TypeCanonicalUser {
+			headers = append(headers, fmt.Sprintf(HEADER_USER_ID_FORMAT, *grant.Grantee.ID))
+		}
+	}
+	return strings.Join(headers, ",")
+}
+
+// GetHeadersFromGrants will return a list of grant headers from grants
+func GetHeadersFromGrants(
```

Comment:
any reason to export this function name? Same for `GetPossibleCannedACLsFromGrants`?

Were you just planning ahead for unit tests using a `bucket_test` package or something like that?

Reply:
No, but now I will because that's a great idea.